projects/igniteui-angular/src/lib/grids/grid-mrl-navigation.service.ts

```diff
@@ -0,0 +1,466 @@
+import { Injectable } from '@angular/core';
+import { IgxGridBaseComponent, FilterMode } from './grid-base.component';
+import { first } from 'rxjs/operators';
+import { IgxColumnComponent } from './column.component';
+import { IgxGridGroupByRowComponent } from './grid/groupby-row.component';
+import { IgxGridCellComponent } from './tree-grid';
+import { IgxGridNavigationService } from './grid-navigation.service';
+
+/** @hidden */
+@Injectable()
+export class IgxGridMRLNavigationService extends IgxGridNavigationService {
+    public grid: IgxGridBaseComponent;
+
+    public navigateUp(rowElement, currentRowIndex, visibleColumnIndex, cell?) {
+        if (cell) {
+            this.focusCellUpFromLayout(cell);
+        } else {
+            super.navigateUp(rowElement, currentRowIndex, visibleColumnIndex);
+        }
+    }
+
+    public navigateDown(rowElement, currentRowIndex, visibleColumnIndex, cell?) {
+        if (cell) {
+            this.focusCellDownFromLayout(cell);
+        } else {
+            super.navigateDown(rowElement, currentRowIndex, visibleColumnIndex);
+        }
+
+    }
+
+    public isColumnFullyVisible(visibleColumnIndex: number) {
+        const forOfDir =  this.grid.headerContainer;
+        const horizontalScroll = forOfDir.getHorizontalScroll();
+        const column = this.grid.columnList.filter(c => !c.columnGroup).find((col) => col.visibleIndex === visibleColumnIndex);
+        if (!horizontalScroll.clientWidth || (column && column.pinned)) {
+            return true;
+        } else if (column) {
+            return this.displayContainerWidth >= forOfDir.getColumnScrollLeft(column.parent.visibleIndex) - this.displayContainerScrollLeft;
+        }
+        return false;
+    }
+
+    public isColumnLeftFullyVisible(visibleColumnIndex: number) {
+        const forOfDir = this.grid.headerContainer;
+        const horizontalScroll = forOfDir.getHorizontalScroll();
+        const column = this.grid.columnList.filter(c => !c.columnGroup).find((col) => col.visibleIndex === visibleColumnIndex);
+        if (!horizontalScroll.clientWidth || column.pinned) {
+            return true;
+        }
+        return this.displayContainerScrollLeft <= forOfDir.getColumnScrollLeft(column.parent.visibleIndex);
+    }
+
+    public onKeydownArrowRight(element, rowIndex, visibleColumnIndex, isSummary = false, cell?) {
+        this.focusNextCellFromLayout(cell);
+    }
+
+    public onKeydownArrowLeft(element, rowIndex, visibleColumnIndex, isSummary = false, cell?) {
+        this.focusPrevCellFromLayout(cell);
+    }
+
+    public performTab(currentRowEl, rowIndex, visibleColumnIndex, isSummaryRow = false, cell?) {
```

Comment:
performTab and performShift tab look so similar - have you tried handling them with the same code with a shift flag determining if you search for +1 or -1?

Reply:
@ChronosSF There are similar in the sense that both find a target element, scroll it in view if necessary and then do something specific to the action ( navigateUp or navigateDown, moveNextEditable or movePreviousEditable etc.).  We cannot handle them completely with the same code without some branching. If you prefer we can extract the common action (find and scroll in view) in a function (something like _scrollToVisibleIndex).